### PR TITLE
feat: add TSViT model architecture

### LINKED ISSUE #4003: New time-series model: TSViT
- Src: ViTs for SITS: Vision Transformers for Satellite Image Time Series" by Tarasiou et al. (2023)
- Motivation: top results on https://huggingface.co/datasets/EOA-team/SwissCrop25
- Implementation: https://github.com/michaeltrs/DeepSatModels/tree/main/models/TSViT
- Checkpoints: https://github.com/michaeltrs/DeepSatModels/blob/main/README_TSVIT.md#pre-trained-checkpoints

@michaeltrs might you be interested in adding? Could also rehost the checkpoints on HuggingFace?

### PR DESCRIPTION
**Pull request by Faissal-00**
### Summary

This PR adds the base TSViT (Time-Series Vision Transformer) architecture.

**What's included:**
* Ported the core mathematical components from the original implementation.
* Formatted to TorchGeo standards with strict typing and docstrings.
* Added basic unit tests for forward pass validation.

**Next Steps:**
I wanted to get the base architecture up for review first. Once this looks good, I will add the logic to pull the pre-trained checkpoints (and we can discuss rehosting them on HuggingFace).

### Rationale

Resolves #4003. 

### Checklist

- [x] I have read the [Contributing docs](https://docs.torchgeo.org/en/stable/user/contributing.html)
- [x] I have read the [AI policy](https://github.com/torchgeo/governance/blob/main/AI-POLICY.md)

- [ ] 🟢 **No AI usage**: written by humans, for humans
- [x] 🟡 **AI-assisted**: AI helped with the coding, but I understand every line
- [ ] 🔴 **AI-generated**: AI did everything, review with caution

### CODE REVIEW

**Comment by Faissal-00** [comment]
Just a quick update: All checks have passed! Ready for review whenever you have the time.

File: torchgeo/models/tsvit.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        temporal_depth: int = 10,
+        spatial_depth: int = 4,
+        heads: int = 3,
+        dim_head: int = 64,
+        dropout: float = 0.0,
+        emb_dropout: float = 0.0,
+        pool: str = 'cls',
+        scale_dim: int = 4,
+    ) -> None:
+        """Initialize the TSViT model."""
```
Comment (nilsleh) [inline torchgeo/models/tsvit.py:67]: needs docstrings, and versionadded parameters, also a reference on where you got the implementation from with appropriate credit and make sure the license was portable

File: torchgeo/models/tsvit.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        self.ff = PreNorm(dim, FeedForward(dim, mlp_dim, dropout=dropout))
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Forward pass."""
+        x = self.attn(x) + x
+        x = self.ff(x) + x
+        return x
+
+
+class Transformer(nn.Module):
```
Comment (nilsleh) [inline torchgeo/models/tsvit.py]: is there somethething special about these transformer blocks and models that could not be simplified with existing parts from timm for example?

**Review by nilsleh — CHANGES_REQUESTED** [review]

File: torchgeo/models/tsvit.py (hunk 1 above)
Comment (Faissal-00) [reply torchgeo/models/tsvit.py:67]: Thanks! I'll add the parameter docstrings and versionadded annotations, along with an explicit reference and credit to the original TSViT implementation and paper. I also checked the source repository: it is licensed under Apache License 2.0, so I'll make sure the required attribution and license notice are handled appropriately for TorchGeo.

File: torchgeo/models/tsvit.py (hunk 2 above)
Comment (Faissal-00) [reply torchgeo/models/tsvit.py]: I initially kept the Attention, PreNorm, FeedForward, and Transformer blocks close to the original implementation to preserve the reference architecture. I’ll check whether the required behavior can be reproduced cleanly using existing timm components and simplify the implementation where appropriate, while preserving the intended TSViT behavior.

File: torchgeo/models/tsvit.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/models/tsvit.py:67]: We cannot include Apache-2.0 code in TorchGeo, TorchGeo is MIT-licensed.

File: torchgeo/models/tsvit.py (hunk 2 above)
Comment (Faissal-00) [reply torchgeo/models/tsvit.py]: @nilsleh I just checked if we could swap these out for timm, but it turns out it won't work here. The paper uses a config with dim=128 and heads=3.
Since 128 isn't divisible by 3, using timm or native PyTorch blocks will crash because they require 
dim % num_heads == 0. Their custom attention block actually bypasses this by projecting the dim to an inner dimension first.
So I think we have to keep the custom blocks so we don't break the math and can still load their pre-trained weights. Let me know if that makes sense!

File: torchgeo/models/tsvit.py (hunk 1 above)
Comment (Faissal-00) [reply torchgeo/models/tsvit.py:67]: @adamjstewart  Understood. I saw the Apache 2.0 license on the original repo but wasn't aware of TorchGeo's strict incompatibility policy against absorbing it.

What is the preferred route for the project here? Should I open an issue on the original author's repository to request explicit MIT relicensing for these specific blocks, or would you prefer a clean-room rewrite of the architecture directly from the paper so the code is entirely original?

Comment (nilsleh) [reply torchgeo/models/tsvit.py:67]: Apache-2 Licensed models can be added in https://github.com/torchgeo/terratorch although you would have to check with the maintainers, since this is not a pretrained EO architecture but just "a" plain time series model. You can also check if they change their license to MIT.

File: torchgeo/models/tsvit.py (hunk 2 above)
Comment (nilsleh) [reply torchgeo/models/tsvit.py]: what is the pretrained checkpoint you are referring to here, is it an EO pretrained model or just some RGB model? 

Currently, you are referencing an arxiv paper that seems to be hallucinated: https://arxiv.org/abs/2301.11942 or at least I fail to understand what it has to do with a TSViT model.

Comment (Faissal-00) [reply torchgeo/models/tsvit.py]: Sorry for the confusion. The checkpoints are EO models trained on the PASTIS dataset (Sentinel-2, 10 bands), which the original authors provided [here in their README](https://github.com/michaeltrs/DeepSatModels/blob/main/README_TSVIT.md#pre-trained-checkpoints). Keeping the custom blocks ensures users can still load these specific weights.

Regarding the arXiv link, I had already updated it locally to arXiv:2301.04944 (ViTs for SITS) and will include it in the next push.

Comment (nilsleh) [reply torchgeo/models/tsvit.py]: My guess is that with the num_heads==3 you are referring to this config: https://github.com/michaeltrs/DeepSatModels/blob/53e655936be088522ec419b3ead38684c20ae1ec/models/TSViT/TSViTcls.py#L109. However, the actual checkpoint that they provide has 4 heads if you inspect it .  Also their model config files seem to suggest 4 heads: https://github.com/michaeltrs/DeepSatModels/blob/main/configs/PASTIS24/TSViT-S_fold1.yaml

So based on caring about that checkpoint, I believe we can actually use the built in timm modules to greatly simplify the implementation.

Comment (nilsleh) [reply torchgeo/models/tsvit.py]:
```
from pathlib import Path

import torch

CHECKPOINT = Path('/path/to/TSViT_fold1/best.pth')

state_dict = torch.load(CHECKPOINT, map_location='cpu', weights_only=True)
dim = state_dict['temporal_token'].shape[-1]
qkv_rows, qkv_columns = state_dict[
    'temporal_transformer.layers.0.0.fn.to_qkv.weight'
].shape
attention_width = qkv_rows // 3

print(f'dim={dim}')
print(f'qkv_weight_shape=({qkv_rows}, {qkv_columns})')
print(f'attention_width={attention_width}')
print(f'3 heads x 64 width would require qkv_weight_shape=(576, {dim})')

assert qkv_columns == dim
assert qkv_rows % 3 == 0
assert attention_width == dim == 128
assert qkv_rows == 384
assert qkv_rows != 3 * 3 * 64
```

dim=128
qkv_weight_shape=(384, 128)
attention_width=128
3 heads x 64 width would require qkv_weight_shape=(576, 128)

File: torchgeo/models/tsvit.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/models/tsvit.py:67]: I wouldn't waste too much time on this PR until we resolve the license question, there is a high probability that this can never be merged into TorchGeo without a clean-room rewrite.

Comment (Faissal-00) [reply torchgeo/models/tsvit.py:67]: Understood. I'll pause changes to the PR until the license question is resolved. If the maintainers decide that a clean-room implementation from the paper is acceptable for TorchGeo, I'd be happy to take that route.

Comment (adamjstewart) [reply torchgeo/models/tsvit.py:67]: Clean-room implementation is definitely acceptable to us, especially if it is able to load pre-trained model weights

**Comment by Faissal-00** [comment]
Addressed the requested changes and pushed an updated implementation:
Reworked TSViT as a clean-room implementation using timm blocks.
Added conversion/loading support for the released PASTIS24 pretrained checkpoint.
Expanded tests to cover validation, checkpoint conversion/loading, and error paths.
Verified numerical agreement with the original implementation using the released TSViT_fold1 checkpoint (allclose=True).

**Comment by nilsleh** [comment]
Hi @Faissal-00 ,

I would have the following general suggestion:
- do the checkpoint conversion offline to clean up the checkpoint once, and then we can upload it to HF https://huggingface.co/torchgeo/models, then we don't have maintain that code and tests also become simpler
- with that hf hosted checkpoint you can then also add automatic download and weight initialization to the TSViT model similarly to how we do this for other models

**Comment by Faissal-00** [comment]
@nilsleh Thanks, that makes sense. I’ve completed the one-time conversion for all five PASTIS24 checkpoints and verified that the converted weights load with `strict=True`.

I don’t appear to have write access to the `torchgeo/models` Hugging Face repository. Could you let me know whether I should request access there, or whether someone from the team should upload the converted checkpoints for me?

**Comment by Faissal-00** [comment]
Hi @nilsleh, just following up on this when you get a chance.

File: torchgeo/models/tsvit.py (hunk 1 above)
Comment (robmarkcole) [reply torchgeo/models/tsvit.py:67]: @Faissal-00 check https://github.com/thomaslauber/SwissCrop25/blob/main/src/tsvit/TSViTdense.py

Actually this appears resolved now

**Comment by nilsleh** [comment]
@adamjstewart what do you prefer regarding the model checkpoint for HF?

**Comment by adamjstewart** [comment]
Not sure I understand the question. The process of adding models to our HF is clear, first add it to your personal HF, then ask the TSC to vote to migrate. Main concern is the license.

**Comment by nilsleh** [comment]
@Faissal-00 I have uploaded the checkpoint here: https://huggingface.co/nilsleh/tsvit/tree/main

you can add automatic downloading capability to the model similar to others in torchgeo with a WeightsEnum like

```python
TSVIT_PASTIS24 = Weights(
      url='https://huggingface.co/nilsleh/tsvit/resolve/061c4c5518629071f7c79d787f0905344d4eb4ab/tsvi
  t_pastis24_fold1-026e8447.pth',
      transforms=...,
      meta={
          'dataset': 'PASTIS24',
          'publication': 'https://arxiv.org/abs/2301.04944',
          'repo': 'https://github.com/michaeltrs/DeepSatModels',
      },
  )
```

**Comment by Faissal-00** [comment]
Hi @nilsleh, updated the TSViT weights integration and tests. Ready for review.

File: torchgeo/models/tsvit.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        Args:
+            x: Input tensor with shape ``(B, T, H, W, C)``.
+
+        Returns:
+            Segmentation logits with shape ``(B, num_classes, H, W)``.
+        """
+        if x.ndim != 5:
+            raise ValueError('Expected input shape (B, T, H, W, C).')
+
+        B, T, H, W, C = x.shape
```
Comment (nilsleh) [inline torchgeo/models/tsvit.py:191]: https://github.com/michaeltrs/DeepSatModels/blob/53e655936be088522ec419b3ead38684c20ae1ec/models/TSViT/TSViTcls.py#L78 

Can you double check the correct ordering of dimensions? It looks like the original code expects B, T, C, H, W?

**Review by nilsleh — CHANGES_REQUESTED** [review]
